tests/test_events.py

```diff
@@ -41,23 +43,64 @@ def setUp(self):
 
     def test_record_event_for_bundle(self):
         metadata_document = dict(foo=f"{uuid4()}")
-        key = f"bundles/{uuid4()}.{datetime_to_version_format(datetime.utcnow())}"
-        test_parameters = [(replica, pfxs) for replica in Replica for pfxs in [None, ("foo", "bar")]]
-        for replica, prefixes in test_parameters:
-            with self.subTest(replica=replica.name, flashflood_prefixes=prefixes):
-                self._test_record_event_for_bundle(replica, prefixes, metadata_document, key)
-
-    def _test_record_event_for_bundle(self, replica, prefixes, metadata_document, key):
-        with mock.patch("dss.events._build_bundle_metadata_document", return_value=metadata_document):
+        keys = [f"bundles/{uuid4()}.{datetime_to_version_format(datetime.utcnow())}",
+                f"bundles/{uuid4()}.{datetime_to_version_format(datetime.utcnow())}.dead",
+                f"bundles/{uuid4()}.dead"]
+        test_parameters = [(replica, key, pfxs)
+                           for replica in Replica
+                           for key in keys
+                           for pfxs in [None, ("foo", "bar")]]
+        for replica, key, prefixes in test_parameters:
+            with self.subTest(replica=replica.name, key=key, flashflood_prefixes=prefixes):
+                if key.endswith(TOMBSTONE_SUFFIX):
+                    with self.assertRaises(events.DSSEventRecordError):
+                        self._test_record_event_for_bundle(replica, key, prefixes, metadata_document)
+                else:
+                    self._test_record_event_for_bundle(replica, key, prefixes, metadata_document)
+                    self._test_record_event_for_bundle(replica, key, prefixes, metadata_document, True)
+
+    def _test_record_event_for_bundle(self, replica, key, prefixes, metadata_document, use_version_for_timestamp=False):
+        with mock.patch("dss.events.build_bundle_metadata_document", return_value=metadata_document):
             with mock.patch("dss.events.FlashFlood") as ff:
-                ret = events.record_event_for_bundle(replica, key, prefixes)
+                ret = events.record_event_for_bundle(replica, key, prefixes, use_version_for_timestamp)
                 used_prefixes = prefixes or replica.flashflood_prefix_write
                 self.assertEqual(len(used_prefixes), ff.call_count)
                 self.assertEqual(metadata_document, ret)
                 for args, pfx in zip(ff.call_args_list, used_prefixes):
                     expected = ((resources.s3, Config.get_flashflood_bucket(), pfx),)
                     self.assertEqual(args, expected)
 
+    def test_delete_event_for_bundle(self):
+        keys = [f"bundles/{uuid4()}.{datetime_to_version_format(datetime.utcnow())}",
+                f"bundles/{uuid4()}.{datetime_to_version_format(datetime.utcnow())}.dead",
+                f"bundles/{uuid4()}.dead"]
+        test_parameters = [(replica, key, pfxs)
```

Comment:
could you use itertools.product here?

Reply:
Possibly, but I'm not sure it would aid readability.